### PR TITLE
Add option to exercises that reveal model solution after deadline or …

### LINKED ISSUE #1145: Enable “instant model solutions” on full-points only
Please correct me if this is outdated/wrong: 

It’s currently possible to configure an exercise so that the student sees the model solution as soon as they score full points **or** run out of attempts. However, it’s not possible to configure an exercise so that the students sees the model solution as soon as they score full points **but not** if they run out of attempts. 

I expect that the former option is quite useful in some courses. However, the latter config option would be much more useful to me (and in some other courses I know). This is because in my context, it’s pretty common for teachers/TAs to add submission attempts, so reaching one’s last attempt is not quite "100% Game Over" yet.

### PR DESCRIPTION
**Pull request by Mikael-Lenander**
…full points

# Description

**What?**

Adds the possibility to configure an exercise so that the model solution is shown after a student gets full points or after the deadline/personal extensions have passed.

**Why?**

O1 (and other courses) need.

**How?**

Adds a new option to RevealRule model.

Related changes in other repositories:
https://github.com/apluslms/gitmanager/pull/41
https://github.com/apluslms/a-plus-rst-tools/pull/167

Fixes #1145

# Testing

**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [x] Django unit tests.
- [ ] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [x] Manual testing.

With this option, model solution is shown after deadline or with full points, but not with all submissions used. Deadline options still works after refactoring.

**Did you test the changes in**

- [x] Chrome
- [x] Firefox
- [ ] This pull request cannot be tested in the browser.

# Translation

- [x] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [x] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?
- [x] Did you use Python type hinting in all functions that you added or edited? ([type hints](https://docs.python.org/3/library/typing.html) for function parameters and return values)

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

### CODE REVIEW

File: exercise/reveal_models.py [hunk 1]
```diff
@@ -54,24 +55,21 @@ def is_revealed(self, state: 'BaseRevealState', time: Optional[datetime.datetime
             RevealRule.TRIGGER.DEADLINE,
             RevealRule.TRIGGER.DEADLINE_ALL
```
Comment (ihalaij1) [inline exercise/reveal_models.py:56]: Could you add a comma at the end of this list? :)

File: exercise/reveal_models.py [hunk 2]
```diff
@@ -54,24 +55,21 @@ def is_revealed(self, state: 'BaseRevealState', time: Optional[datetime.datetime
...
         ]:
-            if time is None:
-                time = timezone.now()
-            reveal_time = self.get_reveal_time(state)
-            if reveal_time is not None:
-                return time > reveal_time
+            if self._is_revealed_due_to_time(state, time):
+                return True
```
Comment (ihalaij1) [inline exercise/reveal_models.py:59]: You can shorten this to ``return self._is_revealed_due_to_time(state, time)``

File: exercise/reveal_models.py [hunk 3]
```diff
@@ -54,24 +55,21 @@ def is_revealed(self, state: 'BaseRevealState', time: Optional[datetime.datetime
...
         elif self.trigger == RevealRule.TRIGGER.COMPLETION:
-            points = state.get_points()
-            max_points = state.get_max_points()
-            if points is not None and max_points is not None:
-                if points >= max_points:
-                    return True
+            if self._is_revealed_due_to_full_points(state):
+                return True
```
Comment (ihalaij1) [inline exercise/reveal_models.py:62]: You can shorten this to ``return self._is_revealed_due_to_full_points(state)``

File: exercise/reveal_models.py [hunk 4]
```diff
@@ -54,24 +55,21 @@ def is_revealed(self, state: 'BaseRevealState', time: Optional[datetime.datetime
...
             submissions = state.get_submissions()
             max_submissions = state.get_max_submissions()
             if submissions is not None and max_submissions is not None:
                 if max_submissions == 0:
                     return False
                 if submissions >= max_submissions:
                     return True
```
Comment (ihalaij1) [inline exercise/reveal_models.py:69]: Could you create a helper function ``_is_revealed_due_to_max_submissions`` for this as well?

File: exercise/reveal_models.py [hunk 5]
```diff
@@ -54,24 +55,21 @@ def is_revealed(self, state: 'BaseRevealState', time: Optional[datetime.datetime
...
+        elif self.trigger == RevealRule.TRIGGER.DEADLINE_OR_FULL_POINTS:
+            if self._is_revealed_due_to_full_points(state) or self._is_revealed_due_to_time(state, time):
+                return True
```
Comment (ihalaij1) [inline exercise/reveal_models.py:72]: You can shorten this to ``return self._is_revealed_due_to_full_points(state) or self._is_revealed_due_to_time(state, time)``

**Review by ihalaij1 — CHANGES_REQUESTED** [review]
Excellent!

I tested this and it seems to work great!
See below for some minor changes.

Also, the commit message should be shorter and include ``Fixes #1145`` on separate line, with an additional empty line above it.

**Comment by Mikael-Lenander** [comment]
Added a delay field to the reveal rule form with the DEADLINE_OR_FULL_POINTS trigger selected.

**Review by ihalaij1 — APPROVED** [review]
Looks good!

**Review by ihalaij1 — CHANGES_REQUESTED** [review]
There's a linter error that should be fixed still.
```
exercise/reveal_models.py
  Line: 49
    pylint: no-else-return / Unnecessary "elif" after "return", remove the leading "el" from "elif"
```

**Comment by Mikael-Lenander** [comment]
Done

**Review by ihalaij1 — APPROVED** [review]
Ready for merge!

**Comment by Mikael-Lenander** [comment]
Still added support to configure this options from files coming from gitmanager

File: exercise/reveal_models.py [hunk 6]
```diff
@@ -94,3 +82,19 @@ def get_reveal_time(self, state: 'BaseRevealState') -> Optional[datetime.datetim
                 return latest_deadline + datetime.timedelta(minutes=self.delay_minutes or 0)
 
         return None
+
+    def _is_revealed_due_to_full_points(self, state: 'BaseRevealState') -> bool:
+        points = state.get_points()
+        max_points = state.get_max_points()
+        return points is not None and max_points is not None and points >= max_points
+
+    def _is_revealed_due_to_time(self, state: 'BaseRevealState', time: datetime.datetime) -> bool:
```
Comment (Mankro) [inline exercise/reveal_models.py:91]: The type hint for `time` is wrong. Since it can be `None`, the type is `Optional[datetime.datetime]`

File: exercise/reveal_models.py [hunk 7]
```diff
@@ -94,3 +82,19 @@ def get_reveal_time(self, state: 'BaseRevealState') -> Optional[datetime.datetim
...
+        if time is None:
+            time = timezone.now()
+        reveal_time = self.get_reveal_time(state)
+        return reveal_time is not None and time > reveal_time
+
+    def _is_revealed_due_to_max_submissions(self, state):
```
Comment (Mankro) [inline exercise/reveal_models.py:97]: Type hints were not added here for the parameters.

**Review by Mankro — APPROVED** [review]
At a quick glance, this looks very nice and clean! @lainets promised to take a closer look and merge this after that.

**Comment by lainets** [comment]
The changes are missing tests and https://github.com/apluslms/a-plus/blob/0d5be74b55ddb352260b39f7b64fc448befa85c2/exercise/viewbase.py#L72 don't include the new option. I'll fix these myself to get this merged quickly.
